### PR TITLE
Update Query Frontend to use Round Robin for Choosing the Next Queue

### LINKED ISSUE #2431: Query Frontend should use Round Robin to Choose the Next Tenant to Service
Currently the query frontend randomly chooses a tenant queue and returns it when a querier requests a new query.  This can cause exaggerated long tail latency if a tenant is simply repeatedly not chosen.

I propose updating the query frontend to use a simple round robin algorithm to choose the next tenant to increase fairness and eliminate the chance of an "unlucky" tenant.

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:
Updates the behavior of the query frontend to use round robin when choosing the next tenant to service.  For lack of a better name a struct "queueManager" was created to provide round robin access to tenant queues.  A combination of a linked list and map were used to provide O(1) for the `deleteQueue`, `getNextQueue` and `getQueue` operations.

**Which issue(s) this PR fixes**:
Fixes #2431 

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by joe-elliott** [comment]
Benchmarks:

**Branch**
```
BenchmarkGetNextRequest-8   	   39643	     38105 ns/op	       0 B/op	       0 allocs/op
BenchmarkQueueRequest-8     	    8538	    143534 ns/op	   56613 B/op	     606 allocs/op

BenchmarkGetNextRequest-8   	   39324	     30886 ns/op	       0 B/op	       0 allocs/op
BenchmarkQueueRequest-8     	   15386	    122518 ns/op	   56606 B/op	     606 allocs/op

BenchmarkGetNextRequest-8   	   33900	     32455 ns/op	       0 B/op	       0 allocs/op
BenchmarkQueueRequest-8     	   12225	    101954 ns/op	   56609 B/op	     606 allocs/op
```

**Master**
```
BenchmarkGetNextRequest-8   	    5682	    233817 ns/op	   53162 B/op	     100 allocs/op
BenchmarkQueueRequest-8     	   12513	    101051 ns/op	   52605 B/op	     506 allocs/op

BenchmarkGetNextRequest-8   	    5647	    190832 ns/op	   53200 B/op	     100 allocs/op
BenchmarkQueueRequest-8     	   16616	     93640 ns/op	   52609 B/op	     506 allocs/op

BenchmarkGetNextRequest-8   	    5505	    225738 ns/op	   53122 B/op	     100 allocs/op
BenchmarkQueueRequest-8     	   12760	     91275 ns/op	   52606 B/op	     506 allocs/op
```

**Comment by joe-elliott** [comment]
@tomwilkie if you could review when you get a chance, i would appreciate it.

File: pkg/querier/frontend/frontend_queues.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import (
+	"container/list"
+)
+
+type queueRecord struct {
+	ch     chan *request
+	userID string
+}
+
+type queueManager struct {
```
Comment (pstibrany) [inline pkg/querier/frontend/frontend_queues.go:12]: Short comment about this struct would be nice.

File: pkg/querier/frontend/frontend_queues_test.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+	// add queues
+	qOne := getOrAddQueue(t, m, "one")
+
+	confirmOrder(t, m, qOne, qOne)
+
+	qTwo := getOrAddQueue(t, m, "two")
+	assert.NotEqual(t, qOne, qTwo)
+
+	confirmOrder(t, m, qOne, qTwo, qOne)
```
Comment (pstibrany) [inline pkg/querier/frontend/frontend_queues_test.go:27]: Very nice tests! They help in understanding how this works.

File: pkg/querier/frontend/frontend_queues.go (hunk 1 above)
Comment (joe-elliott) [reply pkg/querier/frontend/frontend_queues.go]: Added a comment and went with `queueIterator`.  Let me know if you'd like any other changes.

Thanks for the review!

**Pull request merged**